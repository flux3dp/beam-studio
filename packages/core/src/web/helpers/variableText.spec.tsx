const mockBeamboxPreferenceRead = jest.fn();

jest.mock('@core/app/actions/beambox/beambox-preference', () => ({
  read: mockBeamboxPreferenceRead,
}));

const insertSvgElement = (svg: SVGElement) => {
  const container = document.createElement('mock-use');

  container.appendChild(svg);
  document.body.appendChild(container);

  return container;
};
const mockImportBarcodeSvgElement = jest.fn().mockImplementation(insertSvgElement);
const mockImportQrCodeSvgElement = jest.fn().mockImplementation(insertSvgElement);

jest.mock('@core/app/components/dialogs/CodeGenerator/svgOperation', () => ({
  importBarcodeSvgElement: mockImportBarcodeSvgElement,
  importQrCodeSvgElement: mockImportQrCodeSvgElement,
}));

const mockGetState = jest.fn();

jest.mock('@core/app/stores/variableText', () => ({
  useVariableTextState: { getState: mockGetState },
}));

const mockAddSubCommand = jest.fn();
const mockUnapply = jest.fn();
const mockBatchCommand = jest
  .fn()
  .mockImplementation(() => ({ addSubCommand: mockAddSubCommand, unapply: mockUnapply }));
const mockChangeTextCommand = jest.fn();
const mockChangeElementCommand = jest.fn();
const mockMoveElementCommand = jest.fn();
const mockRemoveElementCommand = jest.fn();

jest.mock('@core/app/svgedit/history/history', () => ({
  BatchCommand: mockBatchCommand,
  ChangeElementCommand: mockChangeElementCommand,
  ChangeTextCommand: mockChangeTextCommand,
  MoveElementCommand: mockMoveElementCommand,
  RemoveElementCommand: mockRemoveElementCommand,
}));

const mockAddCommandToHistory = jest.fn();
const mockBeginUndoableChange = jest.fn();
const mockFinishUndoableChange = jest.fn();

jest.mock('@core/app/svgedit/history/undoManager', () => ({
  addCommandToHistory: mockAddCommandToHistory,
  beginUndoableChange: mockBeginUndoableChange,
  finishUndoableChange: mockFinishUndoableChange,
}));

const mockClear = jest.fn();

jest.mock('@core/app/svgedit/text/textactions', () => ({
  clear: mockClear,
}));

const mockRenderText = jest.fn().mockImplementation((elem, value) => {
  elem.innerHTML = `<tspan>${value.replaceAll('\u0085', '</tspan><tspan>')}</tspan>`;
});

jest.mock('@core/app/svgedit/text/textedit', () => ({
  renderText: mockRenderText,
}));

jest.mock('@core/helpers/layer/layer-helper', () => ({
  getObjectLayer: () => 'mock-layer',
}));

const mockClearSelection = jest.fn();

jest.mock('@core/helpers/svg-editor-helper', () => ({
  getSVGAsync: (callback) =>
    callback({
      Canvas: {
        clearSelection: mockClearSelection,
        getCurrentDrawing: () => ({ setCurrentLayer: jest.fn() }),
        selectorManager: { releaseSelectors: jest.fn() },
      },
    }),
}));

const mockBody = {
  convertVariableText: `
<g id="svg_1" class="layer">
<text id="svg_2"><tspan>ddd</tspan></text>
<text id="svg_3" data-vt-type="1" data-vt-offset="1234"><tspan>ddd</tspan></text>
<text id="svg_4" data-vt-type="1" data-vt-offset="30"><tspan>ddd</tspan></text>
<text id="svg_5" data-vt-type="1" data-vt-offset="30"><tspan>0ddd</tspan></text>
<text id="svg_6" data-vt-type="1" data-vt-offset="30"><tspan>0hhh</tspan></text>
<text id="svg_7" data-vt-type="1" data-vt-offset="30"><tspan>0HHH</tspan></text>
<text id="svg_8" data-vt-type="1" data-vt-offset="30"><tspan>Not</tspan><tspan>Number</tspan></text>
<text id="svg_9" data-vt-type="2"><tspan>[YYYYescape] YYYY-MM-DD</tspan><tspan>HH:mm:ss.SSSZ[Z]</tspan><tspan>dddd MMMM</tspan></text>
<text id="svg_10" data-vt-type="3" data-vt-offset="0"><tspan>Name: %0</tspan><tspan>Type: %3</tspan><tspan>Key: %1</tspan></text>
<text id="svg_11" data-vt-type="3" data-vt-offset="1"><tspan>Name: %0</tspan><tspan>Type: %3</tspan><tspan>Key: %1</tspan></text>
<text id="svg_12" data-vt-type="3" data-vt-offset="2"><tspan>Name: %0</tspan><tspan>Type: %3</tspan><tspan>Key: %1</tspan></text>
<use id="svg_13" data-vt-type="1" data-vt-offset="0" data-invert="false" data-props="{&quot;errorLevel&quot;:&quot;L&quot;,&quot;isInvert&quot;:false,&quot;value&quot;:&quot;ddd&quot;}" data-code="qrcode"></use>
<use id="svg_14" data-vt-type="2" data-vt-offset="0" data-invert="true" data-props="{&quot;options&quot;:{&quot;background&quot;:&quot;#000000&quot;,&quot;displayValue&quot;:true,&quot;ean128&quot;:false,&quot;font&quot;:&quot;Noto Sans&quot;,&quot;fontOptions&quot;:&quot; bold&quot;,&quot;fontSize&quot;:20,&quot;format&quot;:&quot;CODE128&quot;,&quot;height&quot;:50,&quot;lineColor&quot;:&quot;#ffffff&quot;,&quot;margin&quot;:10,&quot;textAlign&quot;:&quot;center&quot;,&quot;textMargin&quot;:2,&quot;textPosition&quot;:&quot;bottom&quot;,&quot;width&quot;:2},&quot;value&quot;:&quot;HH:mm:ss&quot;}" data-code="barcode"></use>
</g>
`,
  extractVariableText: `
<g id="svgcontent">
<g id="svg_1" class="layer">
<text id="svg_2"><tspan></tspan></text>
<text id="svg_3" data-vt-type="1"><tspan></tspan></text>
<text id="svg_4" data-vt-type="1"><tspan></tspan></text>
</g>
<g id="svg_5" class="layer">
<text id="svg_6"><tspan></tspan></text>
</g>
<g id="svg_7" class="layer" display="none">
<text id="svg_8" data-vt-type="1"></text>
</g>
</g>
`,
  getVariableTexts: `
<g id="svg_1" class="layer" display="none">
<text id="svg_2" ></text>
<text id="svg_3" data-vt-type="0"></text>
<text id="svg_4" data-vt-type="1"></text>
<text id="svg_5" data-vt-type="2"></text>
<text id="svg_6" data-vt-type="3"></text>
</g>
<g id="svg_11" class="layer">
<text id="svg_7" data-vt-type="0"></text>
<text id="svg_8" data-vt-type="1"></text>
<text id="svg_9" data-vt-type="2"></text>
<text id="svg_10" data-vt-type="3"></text>
</g>
`,
  removeVariableText: `
<g id="svg_1" class="layer">
<text id="svg_2"></text>
<text id="svg_3" data-vt-type="1"></text>
<text id="svg_4" data-vt-type="1"></text>
</g>
<g id="svg_5" class="layer">
<text id="svg_6"></text>
<text id="svg_7" data-vt-type="1"></text>
</g>
<g id="svg_8" class="layer" display="none">
<text id="svg_9" data-vt-type="1"></text>
</g>
`,
};

import { VariableTextType } from '@core/interfaces/ObjectPanel';
import {
  convertVariableText,
  extractVariableText,
  getRealCurrent,
  getVariableTextOffset,
  getVariableTexts,
  getVariableTextType,
  hasVariableText,
  isVariableTextSupported,
  removeVariableText,
} from './variableText';

describe('test variableText helper', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    mockBeamboxPreferenceRead.mockReturnValue('fpm1');
  });

  test('isVariableTextSupported', () => {
    expect(isVariableTextSupported()).toBe(true);
  });

  test('getVariableTextType', () => {
    let elem: SVGTextElement;

    document.body.innerHTML = '<text></text>';
    elem = document.querySelector('text');
    expect(getVariableTextType(elem)).toBe(VariableTextType.NONE);

    document.body.innerHTML = '<text data-vt-type="0"></text>';
    elem = document.querySelector('text');
    expect(getVariableTextType(elem)).toBe(VariableTextType.NONE);

    document.body.innerHTML = '<text data-vt-type="1"></text>';
    elem = document.querySelector('text');
    expect(getVariableTextType(elem)).toBe(VariableTextType.NUMBER);

    document.body.innerHTML = '<text data-vt-type="2"></text>';
    elem = document.querySelector('text');
    expect(getVariableTextType(elem)).toBe(VariableTextType.TIME);

    document.body.innerHTML = '<text data-vt-type="3"></text>';
    elem = document.querySelector('text');
    expect(getVariableTextType(elem)).toBe(VariableTextType.CSV);
  });

  test('getVariableTextOffset', () => {
    let elem: SVGTextElement;

    document.body.innerHTML = '<text></text>';
    elem = document.querySelector('text');
    expect(getVariableTextOffset(elem)).toBe(0);

    document.body.innerHTML = '<text data-vt-offset="0"></text>';
    elem = document.querySelector('text');
    expect(getVariableTextOffset(elem)).toBe(0);

    document.body.innerHTML = '<text data-vt-offset="10"></text>';
    elem = document.querySelector('text');
    expect(getVariableTextOffset(elem)).toBe(10);
  });

  test('getVariableTexts', () => {
    document.body.innerHTML = mockBody.getVariableTexts;

    const checkIds = (elems: NodeListOf<SVGElement>, ids: string[]) => {
      expect(Array.from(elems).map((elem) => elem.id)).toEqual(ids);
    };

    checkIds(getVariableTexts(), ['svg_4', 'svg_5', 'svg_6', 'svg_8', 'svg_9', 'svg_10']);
    checkIds(getVariableTexts({ visibleOnly: true }), ['svg_8', 'svg_9', 'svg_10']);
    checkIds(getVariableTexts({ type: VariableTextType.NUMBER }), ['svg_4', 'svg_8']);
    checkIds(getVariableTexts({ type: VariableTextType.NUMBER, visibleOnly: true }), ['svg_8']);
    checkIds(getVariableTexts({ type: VariableTextType.TIME }), ['svg_5', 'svg_9']);
    checkIds(getVariableTexts({ type: VariableTextType.TIME, visibleOnly: true }), ['svg_9']);
    checkIds(getVariableTexts({ type: VariableTextType.CSV }), ['svg_6', 'svg_10']);
    checkIds(getVariableTexts({ type: VariableTextType.CSV, visibleOnly: true }), ['svg_10']);
  });

  test('isVariableTextExist', () => {
    document.body.innerHTML = '<text data-vt-type="0"></text>';
    expect(hasVariableText()).toBe(false);

    document.body.innerHTML = '<g class="layer" display="none"><text data-vt-type="1"></text></g>';
    expect(hasVariableText()).toBe(true);
    expect(hasVariableText({ visibleOnly: true })).toBe(false);

    document.body.innerHTML = '<text data-vt-type="1"></text><g></g>';
    expect(hasVariableText({ root: document.querySelector('g') })).toBe(false);
  });

  test('getRealCurrent', () => {
    expect(getRealCurrent(0, 0, 9)).toBe(0);
    expect(getRealCurrent(5, 0, 9)).toBe(5);
    expect(getRealCurrent(9, 0, 9)).toBe(9);
    expect(getRealCurrent(15, 0, 9)).toBe(5);
    expect(getRealCurrent(5, 10, 19)).toBe(15);
    expect(getRealCurrent(5, 19, 19)).toBe(19);
    expect(getRealCurrent(5, 30, 19)).toBe(30);
  });

  test('convertVariableText', async () => {
    let revert: (() => void) | null;
    let expectedResult: HTMLElement;
    const mockState: any = {
      csvContent: [
        ['Name 1', 'Key 1', 'Not important', 'Type 1'],
        ['Name 2', 'Key 2'],
      ],
      current: 0,
      end: 999,
      start: 0,
    };

    jest.useFakeTimers();
    jest.setSystemTime(new Date('2025-02-03T12:34:56.1234'));
    Object.defineProperty(navigator, 'language', {
      value: 'zh-TW',
    });
    jest.spyOn(Intl.DateTimeFormat.prototype, 'resolvedOptions').mockReturnValue({
      calendar: 'gregory',
      locale: 'zh-TW',
      numberingSystem: 'latn',
      timeZone: 'Asia/Taipei',
    });
    mockGetState.mockReturnValue(mockState);

    document.body.innerHTML = mockBody.convertVariableText;
    revert = await convertVariableText();
    expect(revert).not.toBe(null);
    expect(mockGetState).toHaveBeenCalled();
    expect(mockClearSelection).toHaveBeenCalled();
    expect(mockClear).toHaveBeenCalled();
    expect(mockChangeTextCommand).toHaveBeenCalledTimes(9);
    expect(mockRemoveElementCommand).toHaveBeenCalledTimes(2);
    expect(mockMoveElementCommand).toHaveBeenCalledTimes(2);
    expect(mockAddSubCommand).toHaveBeenCalledTimes(13);
    expect(mockRenderText).toHaveBeenCalledTimes(9);
    expect(mockImportBarcodeSvgElement).toHaveBeenCalledTimes(1);
    expect(mockImportQrCodeSvgElement).toHaveBeenCalledTimes(1);
    expect(mockChangeElementCommand).not.toHaveBeenCalled();
    expect(mockAddCommandToHistory).not.toHaveBeenCalled();
    expect(document.body).toMatchSnapshot();
    expectedResult = document.body;

    revert();
    expect(mockUnapply).toHaveBeenCalledTimes(1);
    expect(mockUnapply).toHaveBeenCalledWith({ handleHistoryEvent: expect.any(Function), renderText: mockRenderText });

    jest.clearAllMocks();
    document.body.innerHTML = mockBody.convertVariableText;
    revert = await convertVariableText({ configs: mockState });
    expect(revert).not.toBe(null);
    expect(mockGetState).not.toHaveBeenCalled();
    expect(mockClearSelection).toHaveBeenCalled();
    expect(mockClear).toHaveBeenCalled();
    expect(mockChangeTextCommand).toHaveBeenCalledTimes(9);
    expect(mockRemoveElementCommand).toHaveBeenCalledTimes(2);
    expect(mockMoveElementCommand).toHaveBeenCalledTimes(2);
    expect(mockAddSubCommand).toHaveBeenCalledTimes(13);
    expect(mockRenderText).toHaveBeenCalledTimes(9);
    expect(mockImportBarcodeSvgElement).toHaveBeenCalledTimes(1);
    expect(mockImportQrCodeSvgElement).toHaveBeenCalledTimes(1);
    expect(mockChangeElementCommand).not.toHaveBeenCalled();
    expect(mockAddCommandToHistory).not.toHaveBeenCalled();
    expect(document.body).toBe(expectedResult);

    revert();
    expect(mockUnapply).toHaveBeenCalledTimes(1);
    expect(mockUnapply).toHaveBeenCalledWith({ handleHistoryEvent: expect.any(Function), renderText: mockRenderText });

    jest.clearAllMocks();
    document.body.innerHTML = mockBody.convertVariableText;
    revert = await convertVariableText({ addToHistory: true, configs: mockState });
    expect(revert).toBe(null);
    expect(mockGetState).not.toHaveBeenCalled();
    expect(mockClearSelection).toHaveBeenCalled();
    expect(mockClear).toHaveBeenCalled();
    expect(mockChangeTextCommand).toHaveBeenCalledTimes(9);
    expect(mockRemoveElementCommand).toHaveBeenCalledTimes(2);
    expect(mockMoveElementCommand).toHaveBeenCalledTimes(2);
    expect(mockRenderText).toHaveBeenCalledTimes(9);
    expect(mockImportBarcodeSvgElement).toHaveBeenCalledTimes(1);
    expect(mockImportQrCodeSvgElement).toHaveBeenCalledTimes(1);
    expect(mockChangeElementCommand).toHaveBeenCalledTimes(10);
    expect(mockAddSubCommand).toHaveBeenCalledTimes(23);
    expect(mockAddCommandToHistory).toHaveBeenCalledTimes(1);
    expect(document.body).toBe(expectedResult);
  });

  test('removeVariableText', () => {
    let revert: (() => void) | null;

    document.body.innerHTML = mockBody.removeVariableText;
    revert = removeVariableText();
    expect(revert).not.toBe(null);
    expect(mockClearSelection).toHaveBeenCalled();
    expect(mockClear).toHaveBeenCalled();
    expect(document.body).toMatchSnapshot();

    revert();
    expect(document.body).toMatchSnapshot();
  });

  test('extractVariableText', () => {
    document.body.innerHTML = mockBody.extractVariableText;

    let expectedExtractResult: HTMLElement;
    let expectedRevertResult: HTMLElement;
    let extractFn = extractVariableText();

    expect(extractFn).not.toBe(null);
    expect(mockClearSelection).toHaveBeenCalledTimes(1);
    expect(mockClear).toHaveBeenCalledTimes(1);
    expect(document.body).toMatchSnapshot();
    expectedExtractResult = document.body;

    extractFn.revert();
    expect(mockClearSelection).toHaveBeenCalledTimes(2);
    expect(mockClear).toHaveBeenCalledTimes(2);
    expect(document.body).toMatchSnapshot();
    expectedRevertResult = document.body;

    extractFn.extract();
    expect(mockClearSelection).toHaveBeenCalledTimes(3);
    expect(mockClear).toHaveBeenCalledTimes(3);
    expect(document.body).toBe(expectedExtractResult);

    jest.clearAllMocks();
    document.body.innerHTML = mockBody.extractVariableText;
    extractFn = extractVariableText(false);
    expect(extractFn).not.toBe(null);
    expect(mockClearSelection).not.toHaveBeenCalled();
    expect(mockClear).not.toHaveBeenCalled();
    expect(document.body.innerHTML).toBe(mockBody.extractVariableText);

    extractFn.extract();
    expect(mockClearSelection).toHaveBeenCalledTimes(1);
    expect(mockClear).toHaveBeenCalledTimes(1);
    expect(document.body).toBe(expectedExtractResult);

    extractFn.revert();
    expect(mockClearSelection).toHaveBeenCalledTimes(2);
    expect(mockClear).toHaveBeenCalledTimes(2);
    expect(document.body).toBe(expectedRevertResult);
  });
});

describe('test variableText helper when not supported', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    mockBeamboxPreferenceRead.mockReturnValue('ado1');
  });

  test('isVariableTextSupported', () => {
    expect(isVariableTextSupported()).toBe(false);
  });

  test('isVariableTextExist', () => {
    document.body.innerHTML = '<text data-vt-type="1"></text>';
    expect(hasVariableText()).toBe(false);
  });

  test('convertVariableText', async () => {
    document.body.innerHTML = mockBody.convertVariableText;
    expect(await convertVariableText()).toBe(null);
    expect(mockClearSelection).not.toHaveBeenCalled();
    expect(mockClear).not.toHaveBeenCalled();
  });

  test('removeVariableText', () => {
    document.body.innerHTML = mockBody.removeVariableText;
    expect(removeVariableText()).toBe(null);
    expect(mockClearSelection).not.toHaveBeenCalled();
    expect(mockClear).not.toHaveBeenCalled();
  });

  test('extractVariableText', () => {
    document.body.innerHTML = mockBody.extractVariableText;
    expect(extractVariableText()).toBe(null);
    expect(mockClearSelection).not.toHaveBeenCalled();
    expect(mockClear).not.toHaveBeenCalled();
  });
});
