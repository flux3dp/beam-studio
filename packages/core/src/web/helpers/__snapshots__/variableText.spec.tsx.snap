// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test variableText helper convertVariableText 1`] = `
<body>
  

  <g
    class="layer"
    id="svg_1"
  >
    

    <text
      id="svg_2"
    >
      <tspan>
        ddd
      </tspan>
    </text>
    

    <text
      data-vt-offset="1234"
      data-vt-type="1"
      id="svg_3"
    >
      <tspan>
        234
      </tspan>
    </text>
    

    <text
      data-vt-offset="30"
      data-vt-type="1"
      id="svg_4"
    >
      <tspan>
        30
      </tspan>
    </text>
    

    <text
      data-vt-offset="30"
      data-vt-type="1"
      id="svg_5"
    >
      <tspan>
        030
      </tspan>
    </text>
    

    <text
      data-vt-offset="30"
      data-vt-type="1"
      id="svg_6"
    >
      <tspan>
        01e
      </tspan>
    </text>
    

    <text
      data-vt-offset="30"
      data-vt-type="1"
      id="svg_7"
    >
      <tspan>
        01E
      </tspan>
    </text>
    

    <text
      data-vt-offset="30"
      data-vt-type="1"
      id="svg_8"
    >
      <tspan>
        Not
      </tspan>
      <tspan>
        Number
      </tspan>
    </text>
    

    <text
      data-vt-type="2"
      id="svg_9"
    >
      <tspan>
        YYYYescape 2025-02-03
      </tspan>
      <tspan>
        12:34:56.123+08:00Z
      </tspan>
      <tspan>
        星期一 二月
      </tspan>
    </text>
    

    <text
      data-vt-offset="0"
      data-vt-type="3"
      id="svg_10"
    >
      <tspan>
        Name: Name 1
      </tspan>
      <tspan>
        Type: Type 1
      </tspan>
      <tspan>
        Key: Key 1
      </tspan>
    </text>
    

    <text
      data-vt-offset="1"
      data-vt-type="3"
      id="svg_11"
    >
      <tspan>
        Name: Name 2
      </tspan>
      <tspan>
        Type: 
      </tspan>
      <tspan>
        Key: Key 2
      </tspan>
    </text>
    

    <text
      data-vt-offset="2"
      data-vt-type="3"
      id="svg_12"
    >
      <tspan>
        Name: 
      </tspan>
      <tspan>
        Type: 
      </tspan>
      <tspan>
        Key: 
      </tspan>
    </text>
    

    <mock-use
      data-ratiofixed=""
      transform=""
      x="0"
      y="0"
    >
      <svg
        height="1000"
        role="img"
        viewBox="0 0 21 21"
        width="1000"
      >
        <path
          d="M0,0 h21v21H0z"
          fill="transparent"
          shape-rendering="crispEdges"
        />
        <path
          d="M0 0h7v1H0zM9 0h1v1H9zM14,0 h7v1H14zM0 1h1v1H0zM6 1h1v1H6zM9 1h4v1H9zM14 1h1v1H14zM20,1 h1v1H20zM0 2h1v1H0zM2 2h3v1H2zM6 2h1v1H6zM11 2h2v1H11zM14 2h1v1H14zM16 2h3v1H16zM20,2 h1v1H20zM0 3h1v1H0zM2 3h3v1H2zM6 3h1v1H6zM9 3h2v1H9zM14 3h1v1H14zM16 3h3v1H16zM20,3 h1v1H20zM0 4h1v1H0zM2 4h3v1H2zM6 4h1v1H6zM8 4h1v1H8zM10 4h1v1H10zM12 4h1v1H12zM14 4h1v1H14zM16 4h3v1H16zM20,4 h1v1H20zM0 5h1v1H0zM6 5h1v1H6zM9 5h2v1H9zM12 5h1v1H12zM14 5h1v1H14zM20,5 h1v1H20zM0 6h7v1H0zM8 6h1v1H8zM10 6h1v1H10zM12 6h1v1H12zM14,6 h7v1H14zM8 7h1v1H8zM10 7h3v1H10zM2 8h2v1H2zM6 8h4v1H6zM13 8h2v1H13zM16 8h1v1H16zM0 9h1v1H0zM2 9h1v1H2zM5 9h1v1H5zM7 9h3v1H7zM12 9h1v1H12zM14 9h1v1H14zM17 9h1v1H17zM1 10h1v1H1zM3 10h2v1H3zM6 10h1v1H6zM9 10h1v1H9zM11 10h2v1H11zM14 10h1v1H14zM17 10h2v1H17zM20,10 h1v1H20zM0 11h1v1H0zM7 11h2v1H7zM10 11h1v1H10zM12 11h1v1H12zM14 11h6v1H14zM0 12h1v1H0zM2 12h1v1H2zM4 12h1v1H4zM6 12h5v1H6zM13 12h3v1H13zM18 12h1v1H18zM8 13h5v1H8zM14 13h1v1H14zM17 13h1v1H17zM20,13 h1v1H20zM0 14h7v1H0zM8 14h3v1H8zM12 14h1v1H12zM14 14h2v1H14zM17 14h1v1H17zM0 15h1v1H0zM6 15h1v1H6zM9 15h2v1H9zM13 15h4v1H13zM18 15h1v1H18zM0 16h1v1H0zM2 16h3v1H2zM6 16h1v1H6zM11 16h3v1H11zM15 16h2v1H15zM20,16 h1v1H20zM0 17h1v1H0zM2 17h3v1H2zM6 17h1v1H6zM8 17h3v1H8zM12 17h8v1H12zM0 18h1v1H0zM2 18h3v1H2zM6 18h1v1H6zM8 18h1v1H8zM10 18h2v1H10zM13 18h1v1H13zM15 18h1v1H15zM0 19h1v1H0zM6 19h1v1H6zM9 19h2v1H9zM12 19h2v1H12zM15 19h1v1H15zM18,19 h3v1H18zM0 20h7v1H0zM9 20h4v1H9zM14 20h1v1H14zM16 20h1v1H16zM19 20h1v1H19z"
          fill="black"
          shape-rendering="crispEdges"
        />
      </svg>
    </mock-use>
    

    <mock-use
      data-ratiofixed=""
      transform=""
      x="0"
      y="0"
    >
      <svg
        class="visible barcode-svg"
        height="92px"
        id="barcode"
        style="transform: translate(0,0)"
        version="1.1"
        viewBox="0 0 266 92"
        width="266px"
        x="0px"
        xmlns="http://www.w3.org/2000/svg"
        y="0px"
      >
        <rect
          height="92"
          style="fill:#000000;"
          width="266"
          x="0"
          y="0"
        />
        <g
          style="fill:#ffffff;"
          transform="translate(10, 10)"
        >
          <rect
            height="50"
            width="4"
            x="0"
            y="0"
          />
          <rect
            height="50"
            width="2"
            x="6"
            y="0"
          />
          <rect
            height="50"
            width="6"
            x="12"
            y="0"
          />
          <rect
            height="50"
            width="2"
            x="22"
            y="0"
          />
          <rect
            height="50"
            width="4"
            x="26"
            y="0"
          />
          <rect
            height="50"
            width="6"
            x="34"
            y="0"
          />
          <rect
            height="50"
            width="6"
            x="44"
            y="0"
          />
          <rect
            height="50"
            width="2"
            x="52"
            y="0"
          />
          <rect
            height="50"
            width="8"
            x="56"
            y="0"
          />
          <rect
            height="50"
            width="6"
            x="66"
            y="0"
          />
          <rect
            height="50"
            width="2"
            x="76"
            y="0"
          />
          <rect
            height="50"
            width="4"
            x="82"
            y="0"
          />
          <rect
            height="50"
            width="4"
            x="88"
            y="0"
          />
          <rect
            height="50"
            width="2"
            x="96"
            y="0"
          />
          <rect
            height="50"
            width="6"
            x="100"
            y="0"
          />
          <rect
            height="50"
            width="4"
            x="110"
            y="0"
          />
          <rect
            height="50"
            width="2"
            x="118"
            y="0"
          />
          <rect
            height="50"
            width="6"
            x="124"
            y="0"
          />
          <rect
            height="50"
            width="6"
            x="132"
            y="0"
          />
          <rect
            height="50"
            width="2"
            x="142"
            y="0"
          />
          <rect
            height="50"
            width="4"
            x="148"
            y="0"
          />
          <rect
            height="50"
            width="4"
            x="154"
            y="0"
          />
          <rect
            height="50"
            width="6"
            x="160"
            y="0"
          />
          <rect
            height="50"
            width="2"
            x="170"
            y="0"
          />
          <rect
            height="50"
            width="4"
            x="176"
            y="0"
          />
          <rect
            height="50"
            width="6"
            x="184"
            y="0"
          />
          <rect
            height="50"
            width="2"
            x="192"
            y="0"
          />
          <rect
            height="50"
            width="4"
            x="198"
            y="0"
          />
          <rect
            height="50"
            width="6"
            x="206"
            y="0"
          />
          <rect
            height="50"
            width="2"
            x="214"
            y="0"
          />
          <rect
            height="50"
            width="4"
            x="220"
            y="0"
          />
          <rect
            height="50"
            width="6"
            x="230"
            y="0"
          />
          <rect
            height="50"
            width="2"
            x="238"
            y="0"
          />
          <rect
            height="50"
            width="4"
            x="242"
            y="0"
          />
          <text
            style="font: bold 20px Noto Sans"
            text-anchor="middle"
            x="123"
            y="72"
          >
            12:34:56
          </text>
        </g>
      </svg>
    </mock-use>
    

  </g>
  

</body>
`;

exports[`test variableText helper extractVariableText 1`] = `
<body>
  

  <g
    id="svgcontent"
  >
    

    <g
      class="layer"
      id="svg_1"
    >
      

      

      <text
        data-vt-type="1"
        id="svg_3"
      >
        <tspan />
      </text>
      

      <text
        data-vt-type="1"
        id="svg_4"
      >
        <tspan />
      </text>
      

    </g>
    

    

    

  </g>
  

</body>
`;

exports[`test variableText helper extractVariableText 2`] = `
<body>
  

  <g
    id="svgcontent"
  >
    

    <g
      class="layer"
      id="svg_1"
    >
      

      <text
        id="svg_2"
      >
        <tspan />
      </text>
      

      <text
        data-vt-type="1"
        id="svg_3"
      >
        <tspan />
      </text>
      

      <text
        data-vt-type="1"
        id="svg_4"
      >
        <tspan />
      </text>
      

    </g>
    

    <g
      class="layer"
      id="svg_5"
    >
      

      <text
        id="svg_6"
      >
        <tspan />
      </text>
      

    </g>
    

    <g
      class="layer"
      display="none"
      id="svg_7"
    >
      

      <text
        data-vt-type="1"
        id="svg_8"
      />
      

    </g>
    

  </g>
  

</body>
`;

exports[`test variableText helper removeVariableText 1`] = `
<body>
  

  <g
    class="layer"
    id="svg_1"
  >
    

    <text
      id="svg_2"
    />
    

    

    

  </g>
  

  <g
    class="layer"
    id="svg_5"
  >
    

    <text
      id="svg_6"
    />
    

    

  </g>
  

  <g
    class="layer"
    display="none"
    id="svg_8"
  >
    

    

  </g>
  

</body>
`;

exports[`test variableText helper removeVariableText 2`] = `
<body>
  

  <g
    class="layer"
    id="svg_1"
  >
    

    <text
      id="svg_2"
    />
    

    <text
      data-vt-type="1"
      id="svg_3"
    />
    

    <text
      data-vt-type="1"
      id="svg_4"
    />
    

  </g>
  

  <g
    class="layer"
    id="svg_5"
  >
    

    <text
      id="svg_6"
    />
    

    <text
      data-vt-type="1"
      id="svg_7"
    />
    

  </g>
  

  <g
    class="layer"
    display="none"
    id="svg_8"
  >
    

    <text
      data-vt-type="1"
      id="svg_9"
    />
    

  </g>
  

</body>
`;
